### PR TITLE
Detect Amazon EC2 instances that use the newer hypervisor

### PR DESCRIPTION
**Pull request by rcoup**
## What does this PR do?

Currently GDAL doesn't detect newer Amazon EC2 instance types (C5, M5, H1, T3) that use the new ["Nitro" hypervisor](https://aws.amazon.com/ec2/faqs/?nc1=h_ls). GDAL looks for `/sys/hypervisor/uuid` that doesn't exist on these instances.

The [AWS docs](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html) list another method for Nitro but it requires root privileges. Although currently undocumented, AWS support confirms that the following DMI identifiers are also supported (and don't require root privileges):

```
/sys/devices/virtual/dmi/id/chassis_vendor:Amazon EC2
/sys/devices/virtual/dmi/id/sys_vendor:Amazon EC2
/sys/devices/virtual/dmi/id/bios_vendor:Amazon EC2
/sys/devices/virtual/dmi/id/chassis_asset_tag:Amazon EC2
/sys/devices/virtual/dmi/id/board_vendor:Amazon EC2
```

Also deprecates the `CPL_AWS_CHECK_HYPERVISOR_UUID` setting and replaces with `CPL_AWS_AUTODETECT_EC2` since I'm sure it will all change again at some point.

New logic for determining whether to hit the network (ie. we're potentially an EC2 instance):
1. if `CPL_AWS_AUTODETECT_EC2` or `CPL_AWS_CHECK_HYPERVISOR_UUID` are set as False, we're an EC2 instance. The default is True and continue:
2. if `/sys/devices/virtual/dmi/id/sys_vendor` exists and is `"Amazon EC2"` then we're an EC2 instance. If the file exists but the string doesn't match, we're not an EC2 instance.
3. if `/sys/hypervisor/uuid` exists and is a UUID of the form `"ec2..."` then we're an EC2 instance. If the file exists but the UUID doesn't match, we're not an EC2 instance.

## Tasklist
 - [x] Check with Ubuntu instances (old + new hypervisors)
 - [x] Check with Debian instance (old + new hypervisors)
 - [x] Check with Amazon Linux 2 instance (old + new hypervisors)
 - [x] Check with Amazon Linux 2018 (CentOS) instance (old + new hypervisors)
 - [ ] ~Add test case(s)~
 - [x] Review
 - [x] Adjust for comments
 - [x] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS: Linux

### CODE REVIEW

File: gdal/port/cpl_aws.cpp [hunk 1]
```diff
@@ -600,35 +600,70 @@ static bool Iso8601ToUnixTime(const char* pszDT, GIntBig* pnUnixTime)
...
+
+    if( ! CPLTestBool(CPLGetConfigOption("CPL_AWS_AUTODETECT_EC2", "YES")) )
     {
-        VSIStatBufL sStat;
-        if( VSIStatL("/sys/hypervisor", &sStat) == 0 )
+        return true;
+    }
+    else
+    {
+        CPLString opt = CPLGetConfigOption("CPL_AWS_CHECK_HYPERVISOR_UUID", nullptr);
```
Comment (rouault) [inline gdal/port/cpl_aws.cpp]: assigning nullptr to CPLString will crash. Use "" as default value for CPLGetConfigOption() and test for !opt.empty()

File: gdal/port/cpl_aws.cpp [hunk 2]
```diff
@@ -600,35 +600,70 @@ static bool Iso8601ToUnixTime(const char* pszDT, GIntBig* pnUnixTime)
...
+                return true;
             }
         }
     }
-    return bAttemptNetworkAccess;
+
+    VSIStatBufL sStat;
+
+    // Nitro Hypervisor instances
+    if( VSIStatL("/sys/devices/virtual/dmi/id/sys_vendor", &sStat) == 0 )
```
Comment (rouault) [inline gdal/port/cpl_aws.cpp]: VSIStatL() test useless here

File: gdal/port/cpl_aws.cpp [hunk 3]
```diff
@@ -600,35 +600,70 @@ static bool Iso8601ToUnixTime(const char* pszDT, GIntBig* pnUnixTime)
...
+        {
+            VSIFReadL( buf, 1, sizeof(buf)-1, fp );
+            bool bMatch = EQUALN( buf, "Amazon EC2", 10 );
+            VSIFCloseL(fp);
+            return bMatch;
+        }
+    }
+
+    // Xen Hypervisor instances
+    if( VSIStatL("/sys/hypervisor/uuid", &sStat) == 0 )
```
Comment (rouault) [inline gdal/port/cpl_aws.cpp]: VSIStatL() test useless here

File: gdal/port/cpl_aws.cpp (hunk 2 above)
Comment (rcoup) [reply gdal/port/cpl_aws.cpp]: Oh, you mean just check whether I can open it and ignore presence/etc? Yeah, that'd work too 👍

Comment (rouault) [reply gdal/port/cpl_aws.cpp]: yes, that will save a system call.

**Comment by rcoup** [comment]
@rouault ok if I merge into 2.3.x as well?

Only change I'd make is maybe dropping the deprecation warning for `CPL_AWS_CHECK_HYPERVISOR_UUID`?

**Comment by rouault** [comment]
> ok if I merge into 2.3.x as well?

yes

**Pull request merged**